packages/fxa-settings/src/lib/oauth/hooks.tsx

```diff
@@ -0,0 +1,229 @@
+  sessionToken: string,
+  keyFetchToken: string,
+  kB: string
+) {
+  if (
+    integration.wantsKeys() &&
+    integration.data.scope &&
+    integration.data.keysJwk &&
+    integration.data.clientId &&
+    sessionToken &&
+    kB &&
+    keyFetchToken
+  ) {
+    const clientKeyData = await authClient.getOAuthScopedKeyData(
+      sessionToken,
+      integration.data.clientId,
+      await integration.getNormalizedScope()
+    );
+
+    if (clientKeyData && Object.keys(clientKeyData).length > 0) {
+      const keys = await createEncryptedBundle(
+        kB,
+        accountUid,
+        clientKeyData,
+        integration.data.keysJwk
+      );
+      return keys;
+    }
+  }
+}
+
+/**
+ * Creates a new OAuth Code
+ * @param sessionToken - The Current session token
+ * @param keysJwe - An encrypted JWE bundle of key material, to be returned to the client when it redeems the authorization code
+ * @returns An OAuth code
+ */
+async function constructOAuthCode(
+  authClient: AuthClient,
+  integration: OAuthIntegration,
+  sessionToken: string,
+  keysJwe: any
+) {
+  const opts: any = {
+    acr_values: integration.data.acrValues,
+    code_challenge: integration.data.codeChallenge,
+    code_challenge_method: integration.data.codeChallengeMethod,
+    scope: await integration.getNormalizedScope(),
+  };
+  if (keysJwe) {
+    opts.keys_jwe = keysJwe;
+  }
+  if (integration.data.accessType === 'offline') {
+    opts.access_type = integration.data.accessType;
+  }
+
+  const result = await authClient.createOAuthCode(
+    sessionToken,
+    integration.data.clientId,
+    integration.data.state,
+    opts
+  );
+
+  if (!result) {
+    throw new OAuthErrorInvalidOAuthCodeResult();
+  }
+
+  return result;
+}
+
+/**
+ * Builds an updated redirect url for the relying party
+ * @param oauthCode
+ * @returns
+ */
+function constructOAuthRedirectUrl(
+  oauthCode: {
+    code: string;
+    state: string;
+  },
+  redirectTo: string
+) {
+  // Update the state of the redirect URI
+  const redirectUri = new URL(redirectTo);
+  if (oauthCode.code) {
+    redirectUri.searchParams.set('code', oauthCode.code);
+  }
+  if (oauthCode.state) {
+    redirectUri.searchParams.set('state', oauthCode.state);
+  }
+  return redirectUri;
+}
+
+export type FinishOAuthFlowHandler = (
+  accountUid: string,
+  sessionToken: string,
+  keyFetchToken: string,
+  unwrapKB: string
+) => Promise<{ redirect: string }>;
+
+type FinishOAuthFlowHandlerResult = {
+  finishOAuthFlowHandler: FinishOAuthFlowHandler;
+  oAuthDataError: null | Error;
+};
+
+/**
+ * Generates a redirect link which relays the new oauth token to the relying party.
+ * @param accountUid - Current account uid
+ * @param sessionToken - Current session token
+ * @param keyFetchToken - Current key fetch token
+ * @param unwrapKB - Used to unwrap the account keys
+ * @returns An object containing the redirect URL, that can relay the new OAuthCode.
+ */
+export function useFinishOAuthFlowHandler(
```

Comment:
Yes - we are likely going to be using that `useAuthClient` hook in our container components for API calls so I figured if we needed to pull it off Context, the container component is the right place to do it. Plus, the way the errors are currently set up for it (with the existing UI), it made sense to handle those "errors" there. I do have that follow up issue filed for error handling there though, I'm fine with us seeing what we feel is best there, though I would like to avoid that `useAuthClient` hook call in the page components.